lib/oli_web/live/objectives/objectives.ex

```diff
@@ -1,253 +1,329 @@
+      build_objectives(project, [], true)
 
-    objective_mappings = ObjectiveEditor.fetch_objective_mappings(project)
+    {:ok,
+      assign(socket,
+        breadcrumbs: [Breadcrumb.new(%{full_title: "Objectives"})],
+        project: project,
+        author: author,
+        objectives: objectives,
+        table_model: table_model,
+        total_count: length(objectives),
+        all_objectives: all_objectives,
+        all_children: all_children
+      )
+    }
+  end
 
-    subscriptions = subscribe(objective_mappings, project.slug)
+  defp build_objectives(project, objectives_attachments, first_load \\ false) do
+    all_objectives =
+      project
+      |> ObjectiveEditor.fetch_objective_mappings()
+      |> Enum.map(& &1.revision)
 
-    objectives_tree = to_objective_tree(objective_mappings)
+    all_children =
+      all_objectives
+      |> Enum.reduce([], fn rev, acc -> rev.children ++ acc end)
+      |> Enum.uniq()
 
-    {:ok,
-     assign(socket,
-       active: :objectives,
-       objective_mappings: objective_mappings,
-       objectives_tree: objectives_tree,
-       breadcrumbs: [Breadcrumb.new(%{full_title: "Objectives"})],
-       changeset: Resources.change_revision(%Revision{}),
-       project: project,
-       subscriptions: subscriptions,
-       attachment_summary: @default_attachment_summary,
-       modal: nil,
-       author: author,
-       force_render: 0,
-       can_delete?: true,
-       edit: :none,
-       title: "Objectives | " <> project.title,
-       parent_slug: ""
-     )}
-  end
+    objectives =
+      Enum.reduce(all_objectives, [], fn rev, acc ->
+        case Enum.find(all_children, fn child_id -> child_id == rev.resource_id end) do
+          nil ->
+            mapped_children = Enum.map(rev.children, fn resource_id ->
+              Enum.find(all_objectives, fn rev -> rev.resource_id == resource_id end)
+            end)
+
+            [Map.merge(rev,
+              %{
+                children: mapped_children,
+                sub_objectives_count: length(mapped_children),
+                page_attachments_count: 0,
+                page_attachments: [],
+                activity_attachments_count: 0,
+              }
+            )] ++ acc
+          _ -> acc
+        end
+      end)
 
-  def render(assigns) do
-    ~L"""
-    <%= render_modal(assigns) %>
-
-    <div class="objectives container">
-      <div class="mb-2 row">
-        <div class="col-12">
-          <p>
-            Learning objectives help you to organize course content and determine appropriate assessments and instructional strategies.
-            <br/>Refer to the <a href="https://www.cmu.edu/teaching/designteach/design/learningobjectives.html" target="_blank">CMU Eberly Center guide on learning objectives</a> to learn more about the importance of attaching learning objectives to pages and activities.
-          </p>
-        </div>
-      </div>
+    pid = self()
+    if first_load do
```

Comment:
We are fetching the attachments just on mount and then use it from the assigns, so we don't make all these queries on every update of objectives. That info is not going to change in this view, and we want to just display it